File: apps/api/app/domain/simulation.py
from decimal import Decimal

from app.domain.decimal_utils import HUNDRED
from app.domain.types import Driver, SimulationResult, Thermometer


def simulate(
    *,
    cost: Decimal | None,
    driver: Driver,
    driver_value: Decimal,
    ideal_percent: Decimal | None = None,
    source_inactive: bool = False,
    source_unknown: bool = False,
) -> SimulationResult:
    warnings: list[str] = []
    if source_inactive:
        warnings.append("inactive_source")
    if source_unknown:
        warnings.append("unknown_source_status")

    if driver is Driver.PRICE and driver_value < 0:
        raise ValueError("Price cannot be negative")

    price: Decimal | None = None
    gain_amount: Decimal | None = None
    gain_percent: Decimal | None = None

    if cost is None:
        warnings.append("missing_cost")
        if driver is Driver.PRICE:
            price = driver_value
        else:
            warnings.append("driver_requires_cost")
    elif cost == 0:
        warnings.append("zero_cost")
        if driver is Driver.PRICE:
            price = driver_value
            gain_amount = price
        elif driver is Driver.GAIN_AMOUNT:
            gain_amount = driver_value
            if gain_amount < 0:
                raise ValueError("Gain cannot produce a negative price")
            price = gain_amount
        else:
            warnings.append("percentage_not_calculable")
    else:
        if driver is Driver.PRICE:
            price = driver_value
            gain_amount = price - cost
            gain_percent = gain_amount / cost * HUNDRED
        elif driver is Driver.GAIN_AMOUNT:
            gain_amount = driver_value
            price = cost + gain_amount
            if price < 0:
                raise ValueError("Gain cannot produce a negative price")
            gain_percent = gain_amount / cost * HUNDRED
        else:
            if driver_value < -HUNDRED:
                raise ValueError("Percentage cannot be lower than -100")
            gain_percent = driver_value
            gain_amount = cost * gain_percent / HUNDRED
            price = cost + gain_amount

    ideal_amount = None
    ideal_price = None
    gap_amount = None
    gap_percentage_points = None
    thermometer = Thermometer.NEUTRAL

    if ideal_percent is None:
        warnings.append("missing_ideal_margin")
    elif cost is None or cost == 0:
        warnings.append("ideal_not_calculable")
    else:
        ideal_amount = cost * ideal_percent / HUNDRED
        ideal_price = cost + ideal_amount
        if gain_amount is not None:
            gap_amount = gain_amount - ideal_amount
        if gain_percent is not None:
            gap_percentage_points = gain_percent - ideal_percent
            thermometer = Thermometer.GREEN if gap_percentage_points >= 0 else Thermometer.RED

    return SimulationResult(
        price=price,
        gain_amount=gain_amount,
        gain_percent=gain_percent,
        ideal_amount=ideal_amount,
        ideal_price=ideal_price,
        gap_amount=gap_amount,
        gap_percentage_points=gap_percentage_points,
        thermometer=thermometer,
        warnings=tuple(dict.fromkeys(warnings)),
    )

File: apps/web/lib/domain/simulation.ts
// Motor de simulación comercial — autoridad de servidor.
// Puerto fiel de la lógica de dominio: doble vía, estados y termómetro.
// Reglas bloqueadas (AGENTS.md): % sobre costo, sin redondeo, sin float.
import { Decimal, HUNDRED } from "./decimal";
import { EngineError, type SimulationInput, type SimulationResult } from "./types";

const ZERO = new Decimal(0);

/** Elimina duplicados conservando el orden de aparición. */
function unique(items: string[]): string[] {
  return Array.from(new Set(items));
}

export function simulate(input: SimulationInput): SimulationResult {
  const { cost, driver, driverValue } = input;
  const idealPercent = input.idealPercent ?? null;
  const warnings: string[] = [];

  if (input.sourceInactive) warnings.push("inactive_source");
  if (input.sourceUnknown) warnings.push("unknown_source_status");

  if (driver === "price" && driverValue.lessThan(0)) {
    throw new EngineError("price_negative");
  }

  let price: Decimal | null = null;
  let gainAmount: Decimal | null = null;
  let gainPercent: Decimal | null = null;

  if (cost === null) {
    warnings.push("missing_cost");
    if (driver === "price") {
      price = driverValue;
    } else {
      warnings.push("driver_requires_cost");
    }
  } else if (cost.isZero()) {
    warnings.push("zero_cost");
    if (driver === "price") {
      price = driverValue;
      gainAmount = price;
    } else if (driver === "gain_amount") {
      gainAmount = driverValue;
      if (gainAmount.lessThan(0)) throw new EngineError("gain_amount_negative_price");
      price = gainAmount;
    } else {
      warnings.push("percentage_not_calculable");
    }
  } else if (driver === "price") {
    price = driverValue;
    gainAmount = price.minus(cost);
    gainPercent = gainAmount.div(cost).times(HUNDRED);
  } else if (driver === "gain_amount") {
    gainAmount = driverValue;
    price = cost.plus(gainAmount);
    if (price.lessThan(0)) throw new EngineError("gain_amount_negative_price");
    gainPercent = gainAmount.div(cost).times(HUNDRED);
  } else {
    if (driverValue.lessThan(HUNDRED.negated())) throw new EngineError("percent_below_minus_100");
    gainPercent = driverValue;
    gainAmount = cost.times(gainPercent).div(HUNDRED);
    price = cost.plus(gainAmount);
  }

  let idealAmount: Decimal | null = null;
  let idealPrice: Decimal | null = null;
  let gapAmount: Decimal | null = null;
  let gapPercentagePoints: Decimal | null = null;
  let thermometer: SimulationResult["thermometer"] = "neutral";

  if (idealPercent === null) {
    warnings.push("missing_ideal_margin");
  } else if (cost === null || cost.isZero()) {
    warnings.push("ideal_not_calculable");
  } else {
    idealAmount = cost.times(idealPercent).div(HUNDRED);
    idealPrice = cost.plus(idealAmount);
    if (gainAmount !== null) gapAmount = gainAmount.minus(idealAmount);
    if (gainPercent !== null) {
      gapPercentagePoints = gainPercent.minus(idealPercent);
      thermometer = gapPercentagePoints.greaterThanOrEqualTo(ZERO) ? "green" : "red";
    }
  }

  return {
    price,
    gainAmount,
    gainPercent,
    idealAmount,
    idealPrice,
    gapAmount,
    gapPercentagePoints,
    thermometer,
    warnings: unique(warnings),
  };
}
